reco_utils/evaluation/spark_evaluation.py

```diff
@@ -470,3 +479,457 @@ def _get_relevant_items_by_timestamp(
+            self.df_intralist_similarity = (
+                pairs.join(similarity_df, on=["i1", "i2"], how="left")
+                .fillna(0)
+                .filter(F.col("i1") != F.col("i2"))
+                .groupBy(self.col_user)
+                .agg(F.mean(self.sim_col).alias("avg_il_sim"))
+                .select(self.col_user, "avg_il_sim")
+            )
+        return self.df_intralist_similarity
+
+    def user_diversity(self):
+        """Calculate average diversity of recommendations for each user.
+        The metric definition is based on formula (3) in the following reference:
+
+        :Citation:
+
+            Y.C. Zhang, D.Ó. Séaghdha, D. Quercia and T. Jambor, Auralist:
+            introducing serendipity into music recommendation, WSDM 2012
+
+        Returns:
+            pyspark.sql.dataframe.DataFrame: A dataframe with the following columns: col_user, user_diversity.
+        """
+        if self.df_user_diversity is None:
+            self.df_intralist_similarity = self._get_intralist_similarity(self.reco_df)
+            self.df_user_diversity = (
+                self.df_intralist_similarity.withColumn(
+                    "user_diversity", 1 - F.col("avg_il_sim")
+                )
+                .select(self.col_user, "user_diversity")
+                .orderBy(self.col_user)
+            )
+        return self.df_user_diversity
+
+    def diversity(self):
+        """Calculate average diversity of recommendations across all users.
+
+        Returns:
+            float: diversity.
+        """
+        if self.df_diversity is None:
+            self.df_user_diversity = self.user_diversity()
+            self.df_diversity = self.df_user_diversity.agg(
+                {"user_diversity": "mean"}
+            ).first()[0]
+        return self.df_diversity
+
+    # Novelty metrics
+    def item_novelty(self):
+        """Calculate novelty for each item. Novelty is computed as the minus logarithm of
+        (number of interactions with item / total number of interactions). The definition of the metric
+        is based on the following reference using the choice model (eqs. 1 and 6):
+
+        :Citation:
+
+            P. Castells, S. Vargas, and J. Wang, Novelty and diversity metrics for recommender systems:
+            choice, discovery and relevance, ECIR 2011
+
+        Returns:
+            pyspark.sql.dataframe.DataFrame: A dataframe with the following columns: col_item, item_novelty.
+        """
+        if self.df_item_novelty is None:
+            n_records = self.train_df.count()
+            self.df_item_novelty = (
+                self.train_df.groupBy(self.col_item)
+                .count()
+                .withColumn("item_novelty", -F.log2(F.col("count") / n_records))
+                .select(self.col_item, "item_novelty")
+                .orderBy(self.col_item)
+            )
+        return self.df_item_novelty
+
+    def user_novelty(self):
```

Comment:
i'm not sure using the ground model definition in this paper is appropriate here, my interpretation is that is useful to estimate what a diversity-focused model should provide, whereas here we already have a model that has generated recommendations so there's no need to estimate probabilities as they can be calculated explicitly from the set of predictions.

Reply:
I am not sure I understand your point, but like item novelty above, user-relative novelty is calculated on historical interactions. It is basically item novelty conditioned on a specific user. So everything they have in section 4 is measured on historical data. In section 5 however they introduce the recommendations, because they define the overall novelty on the recommendations (which corresponds to `novelty` in the code).

Reply:
see comment above, so far we've been looking at the problem from the perspective of evaluating the recommendations from a model not the historical interactions. if we use this approach comparing results of ALS vs a random recommendation algorithm will be equal, which seems unintuitive to me.

Reply:
They are not equal, see above. What happens that may be a bit confusing is that there is 
- novelty per item (whether plain or user-relative)
- aggregate novelty over the recommendations. 
The former depends only on the historical data, the latter on both historical data and recommendations.

Reply:
where is this used? should there be a flag in novelty to use item or user-relative novelties?

Reply:
There are no flags. There are three class methods:
- `item_novelty` (i.e. novelty per item, across all users)
- `user_novelty` (novelty per item, user-relative)
- `novelty` (aggregate novelty over a set of recommendations).

See the docstrings about which definitions in the paper they implement. The "headline" metric shown in the notebook is `novelty` (but its computation requires `item_novelty`).

Reply:
I get that internally that item novelty (assessed on historical data) is used in calculating novelty. But do we expect users to access the raw item_novelty or user_novelty results? Is there an aggregate form of user_novelty based on recommendations that is useful?

we might want to make it clear they do not use the recommendations, by naming them historical_item_novelty or something similar.

Reply:
`user_novelty` is not referenced anywhere, but Yan had already written a version, so I made it consistent with the paper.

Reply:
Yes, good suggestion, I will rename them. It may be useful to have access to the novelty of a certain item outside recommendations e.g. for data pre or post processing or analytics, although it's probably not a frequent scenario.
There is an aggregate version of user novelty in the definition of `m(R|u)` in section 5 of the paper. Maybe it is good to add it too.

Reply:
I agree that would be nice to add